File: app/src/main/java/info/nightscout/androidaps/plugins/pump/insight/satl/SatlMessage.java
package info.nightscout.androidaps.plugins.pump.insight.satl;

import java.util.Arrays;

import info.nightscout.androidaps.plugins.pump.insight.exceptions.IncompatibleSatlVersionException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidMacTrailerException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidNonceException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidPacketLengthsException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidPreambleException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidSatlCRCException;
import info.nightscout.androidaps.plugins.pump.insight.exceptions.InvalidSatlCommandException;
import info.nightscout.androidaps.plugins.pump.insight.ids.SatlCommandIDs;
import info.nightscout.androidaps.plugins.pump.insight.utils.ByteBuf;
import info.nightscout.androidaps.plugins.pump.insight.utils.Nonce;
import info.nightscout.androidaps.plugins.pump.insight.utils.crypto.Cryptograph;

public abstract class SatlMessage {

    private static final long PREAMBLE = 4293840008L;
    private static final byte VERSION = 0x20;

    private Nonce nonce;
    private long commID = 0;
    private byte[] satlContent;

    protected ByteBuf getData() {
        return new ByteBuf(0);
    }

    protected void parse(ByteBuf byteBuf) {

    }

    public ByteBuf serialize(Class<? extends SatlMessage> clazz, byte[] key) {
        ByteBuf byteBuf;
        if (nonce == null || key == null) byteBuf = serializeCRC(clazz);
        else byteBuf = serializeCTR(nonce.getProductionalBytes(), key, SatlCommandIDs.IDS.getID(clazz));
        satlContent = byteBuf.getBytes(8, byteBuf.getSize() - 16);
        return byteBuf;
    }

    private ByteBuf serializeCRC(Class<? extends SatlMessage> clazz) {
        ByteBuf data = getData();
        int length = data.getSize() + 31;
        ByteBuf byteBuf = new ByteBuf(length + 8);
        byteBuf.putUInt32LE(PREAMBLE);
        byteBuf.putUInt16LE(length);
        byteBuf.putUInt16LE(~length);
        byteBuf.putByte(VERSION);
        byteBuf.putByte(SatlCommandIDs.IDS.getID(clazz));
        byteBuf.putUInt16LE(data.getSize() + 2);
        byteBuf.putUInt32LE(clazz == KeyRequest.class ? 1 : commID);
        byteBuf.putBytes((byte) 0x00, 13);
        byteBuf.putByteBuf(data);
        byteBuf.putUInt16LE(Cryptograph.calculateCRC(byteBuf.getBytes(8, length - 10)));
        byteBuf.putBytes((byte) 0x00, 8);
        return byteBuf;
    }

    private ByteBuf serializeCTR(ByteBuf nonce, byte[] key, byte commandId) {
        ByteBuf data = getData();
        ByteBuf encryptedData = ByteBuf.from(Cryptograph.encryptDataCTR(data.getBytes(), key, nonce.getBytes()));
        int length = 29 + encryptedData.getSize();
        ByteBuf byteBuf = new ByteBuf(length + 8);
        byteBuf.putUInt32LE(PREAMBLE);
        byteBuf.putUInt16LE(length);
        byteBuf.putUInt16LE(~length);
        byteBuf.putByte(VERSION);
        byteBuf.putByte(commandId);
        byteBuf.putUInt16LE(encryptedData.getSize());
        byteBuf.putUInt32LE(commID);
        byteBuf.putByteBuf(nonce);
        byteBuf.putByteBuf(encryptedData);
        byteBuf.putBytes(Cryptograph.produceCCMTag(byteBuf.getBytes(16, 13), data.getBytes(), byteBuf.getBytes(8, 21), key));
        return byteBuf;
    }

    public static SatlMessage deserialize(ByteBuf data, Nonce lastNonce, byte[] key) throws InvalidMacTrailerException, InvalidSatlCRCException, InvalidNonceException, InvalidPreambleException, InvalidPacketLengthsException, IncompatibleSatlVersionException, InvalidSatlCommandException {
        SatlMessage satlMessage;
        byte[] satlContent = data.getBytes(8, data.getSize() - 16);
        if (key == null) satlMessage = deserializeCRC(data);
        else satlMessage = deserializeCTR(data, lastNonce, key);
        satlMessage.setSatlContent(satlContent);
        return satlMessage;
    }

    private static SatlMessage deserializeCTR(ByteBuf data, Nonce lastNonce, byte[] key) throws InvalidMacTrailerException, InvalidNonceException, InvalidPreambleException, InvalidPacketLengthsException, IncompatibleSatlVersionException, InvalidSatlCommandException {
        long preamble = data.readUInt32LE();
        int packetLength = data.readUInt16LE();
        int packetLengthXOR = data.readUInt16LE() ^ 65535;
        byte[] header = data.getBytes(21);
        byte version = data.readByte();
        byte commandId = data.readByte();
        Class<? extends SatlMessage> clazz = SatlCommandIDs.IDS.getType(commandId);
        int dataLength = data.readUInt16LE();
        long commId = data.readUInt32LE();
        byte[] nonce = data.readBytes(13);
        byte[] payload = data.readBytes(dataLength);
        byte[] trailer = data.readBytes(8);
        Nonce parsedNonce = Nonce.fromProductionalBytes(nonce);
        payload = Cryptograph.encryptDataCTR(payload, key, nonce);
        if (!Arrays.equals(trailer, Cryptograph.produceCCMTag(nonce, payload, header, key))) throw new InvalidMacTrailerException();
        if (!lastNonce.isSmallerThan(parsedNonce)) throw new InvalidNonceException();
        if (preamble != PREAMBLE) throw new InvalidPreambleException();
        if (packetLength != packetLengthXOR) throw new InvalidPacketLengthsException();
        if (version != VERSION) throw new IncompatibleSatlVersionException();
        if (clazz == null) throw new InvalidSatlCommandException();
        SatlMessage message = null;
        try {
            message = clazz.newInstance();
        } catch (Exception ignored) {
        }
        message.parse(ByteBuf.from(payload));
        message.setNonce(parsedNonce);
        message.setCommID(commId);
        return message;
    }

    private static SatlMessage deserializeCRC(ByteBuf data) throws InvalidSatlCRCException, InvalidPreambleException, InvalidPacketLengthsException, IncompatibleSatlVersionException, InvalidSatlCommandException {
        long preamble = data.readUInt32LE();
        int packetLength = data.readUInt16LE();
        int packetLengthXOR = data.readUInt16LE() ^ 65535;
        byte[] crcContent = data.getBytes(packetLength - 10);
        byte version = data.readByte();
        byte commandId = data.readByte();
        Class<? extends SatlMessage> clazz = SatlCommandIDs.IDS.getType(commandId);
        int dataLength = data.readUInt16LE();
        long commId = data.readUInt32LE();
        byte[] nonce = data.readBytes(13);
        byte[] payload = data.readBytes(dataLength - 2);
        int crc = data.readUInt16LE();
        data.shift(8);
        if (crc != Cryptograph.calculateCRC(crcContent)) throw new InvalidSatlCRCException();
        if (preamble != PREAMBLE) throw new InvalidPreambleException();
        if (packetLength != packetLengthXOR) throw new InvalidPacketLengthsException();
        if (version != VERSION) throw new IncompatibleSatlVersionException();
        if (clazz == null) throw new InvalidSatlCommandException();
        SatlMessage message = null;
        try {
            message = clazz.newInstance();
        } catch (Exception ignored) {
        }
        message.parse(ByteBuf.from(payload));
        message.setNonce(Nonce.fromProductionalBytes(nonce));
        message.setCommID(commId);
        return message;
    }

    public static boolean hasCompletePacket(ByteBuf byteBuf) {
        if (byteBuf.getSize() < 37) return false;
        if (byteBuf.getSize() < byteBuf.getUInt16LE(4) + 8) return false;
        return true;
    }

    public Nonce getNonce() {
        return this.nonce;
    }

    public long getCommID() {
        return this.commID;
    }

    public byte[] getSatlContent() {
        return this.satlContent;
    }

    public void setNonce(Nonce nonce) {
        this.nonce = nonce;
    }

    public void setCommID(long commID) {
        this.commID = commID;
    }

    public void setSatlContent(byte[] satlContent) {
        this.satlContent = satlContent;
    }
}

File: pump/insight/src/main/kotlin/app/aaps/pump/insight/satl/SatlMessage.kt
package app.aaps.pump.insight.satl

import app.aaps.pump.insight.descriptors.SatlCommands
import app.aaps.pump.insight.exceptions.IncompatibleSatlVersionException
import app.aaps.pump.insight.exceptions.InvalidMacTrailerException
import app.aaps.pump.insight.exceptions.InvalidNonceException
import app.aaps.pump.insight.exceptions.InvalidPacketLengthsException
import app.aaps.pump.insight.exceptions.InvalidPreambleException
import app.aaps.pump.insight.exceptions.InvalidSatlCRCException
import app.aaps.pump.insight.exceptions.InvalidSatlCommandException
import app.aaps.pump.insight.utils.ByteBuf
import app.aaps.pump.insight.utils.Nonce
import app.aaps.pump.insight.utils.crypto.Cryptograph

abstract class SatlMessage {

    var nonce: Nonce? = null
    var commID: Long = 0
    lateinit var satlContent: ByteArray
    protected open val data: ByteBuf
        get() = ByteBuf(0)

    protected open fun parse(byteBuf: ByteBuf) = Unit

    fun serialize(key: ByteArray?): ByteBuf {
        val byteBuf: ByteBuf = if (nonce == null || key == null) serializeCRC() else serializeCTR(nonce!!.productionalBytes, key, SatlCommands.fromType(this))
        satlContent = byteBuf.getBytes(8, byteBuf.filledSize - 16)
        return byteBuf
    }

    private fun serializeCRC(): ByteBuf {
        val length = data.filledSize + 31
        val byteBuf = ByteBuf(length + 8)
        byteBuf.putUInt32LE(PREAMBLE)
        byteBuf.putUInt16LE(length)
        byteBuf.putUInt16LE(length.inv())
        byteBuf.putByte(VERSION)
        byteBuf.putByte(SatlCommands.fromType(this))
        byteBuf.putUInt16LE(data.filledSize + 2)
        byteBuf.putUInt32LE(if (this is KeyRequest) 1 else commID)
        byteBuf.putBytes(0x00.toByte(), 13)
        byteBuf.putByteBuf(data)
        byteBuf.putUInt16LE(Cryptograph.calculateCRC(byteBuf.getBytes(8, length - 10)))
        byteBuf.putBytes(0x00.toByte(), 8)
        return byteBuf
    }

    private fun serializeCTR(nonce: ByteBuf, key: ByteArray, commandId: Byte): ByteBuf {
        val data = data
        val encryptedData = ByteBuf.from(Cryptograph.encryptDataCTR(data.bytes, key, nonce.bytes))
        val length = 29 + encryptedData.filledSize
        val byteBuf = ByteBuf(length + 8)
        byteBuf.putUInt32LE(PREAMBLE)
        byteBuf.putUInt16LE(length)
        byteBuf.putUInt16LE(length.inv())
        byteBuf.putByte(VERSION)
        byteBuf.putByte(commandId)
        byteBuf.putUInt16LE(encryptedData.filledSize)
        byteBuf.putUInt32LE(commID)
        byteBuf.putByteBuf(nonce)
        byteBuf.putByteBuf(encryptedData)
        byteBuf.putBytes(Cryptograph.produceCCMTag(byteBuf.getBytes(16, 13), data.bytes, byteBuf.getBytes(8, 21), key))
        return byteBuf
    }

    companion object {

        private const val PREAMBLE = 4293840008L
        private const val VERSION: Byte = 0x20

        @JvmStatic
        @Throws(
            InvalidMacTrailerException::class,
            InvalidSatlCRCException::class,
            InvalidNonceException::class,
            InvalidPreambleException::class,
            InvalidPacketLengthsException::class,
            IncompatibleSatlVersionException::class,
            InvalidSatlCommandException::class
        )
        fun deserialize(data: ByteBuf, lastNonce: Nonce?, key: ByteArray?): SatlMessage? {
            val satlContent = data.getBytes(8, data.filledSize - 16)
            val satlMessage: SatlMessage? = if (key == null) deserializeCRC(data) else lastNonce?.let { deserializeCTR(data, it, key) }
            satlMessage?.let { it.satlContent = satlContent }
            return satlMessage
        }

        @Throws(
            InvalidMacTrailerException::class,
            InvalidNonceException::class,
            InvalidPreambleException::class,
            InvalidPacketLengthsException::class,
            IncompatibleSatlVersionException::class,
            InvalidSatlCommandException::class
        )
        private fun deserializeCTR(data: ByteBuf, lastNonce: Nonce, key: ByteArray): SatlMessage {
            val preamble = data.readUInt32LE()
            val packetLength = data.readUInt16LE()
            val packetLengthXOR = data.readUInt16LE() xor 65535
            val header = data.getBytes(21)
            val version = data.readByte()
            val commandId = data.readByte()
            val message = SatlCommands.fromId(commandId)
            val dataLength = data.readUInt16LE()
            val commId = data.readUInt32LE()
            val nonce = data.readBytes(13)
            var payload = data.readBytes(dataLength)
            val trailer = data.readBytes(8)
            val parsedNonce = Nonce.fromProductionalBytes(nonce)
            payload = Cryptograph.encryptDataCTR(payload, key, nonce)
            if (!trailer.contentEquals(Cryptograph.produceCCMTag(nonce, payload, header, key))) throw InvalidMacTrailerException()
            if (!lastNonce.isSmallerThan(parsedNonce)) throw InvalidNonceException()
            if (preamble != PREAMBLE) throw InvalidPreambleException()
            if (packetLength != packetLengthXOR) throw InvalidPacketLengthsException()
            if (version != VERSION) throw IncompatibleSatlVersionException()
            if (message == null) throw InvalidSatlCommandException()
            return message.also {
                it.parse(ByteBuf.from(payload))
                it.nonce = parsedNonce
                it.commID = commId
            }
        }

        @Throws(InvalidSatlCRCException::class, InvalidPreambleException::class, InvalidPacketLengthsException::class, IncompatibleSatlVersionException::class, InvalidSatlCommandException::class)
        private fun deserializeCRC(data: ByteBuf): SatlMessage {
            val preamble = data.readUInt32LE()
            val packetLength = data.readUInt16LE()
            val packetLengthXOR = data.readUInt16LE() xor 65535
            val crcContent = data.getBytes(packetLength - 10)
            val version = data.readByte()
            val commandId = data.readByte()
            val message = SatlCommands.fromId(commandId)
            val dataLength = data.readUInt16LE()
            val commId = data.readUInt32LE()
            val nonce = data.readBytes(13)
            val payload = data.readBytes(dataLength - 2)
            val crc = data.readUInt16LE()
            data.shift(8)
            if (crc != Cryptograph.calculateCRC(crcContent)) throw InvalidSatlCRCException()
            if (preamble != PREAMBLE) throw InvalidPreambleException()
            if (packetLength != packetLengthXOR) throw InvalidPacketLengthsException()
            if (version != VERSION) throw IncompatibleSatlVersionException()
            if (message == null) throw InvalidSatlCommandException()
            return message.also {
                it.parse(ByteBuf.from(payload))
                it.nonce = Nonce.fromProductionalBytes(nonce)
                it.commID = commId
            }
        }

        @JvmStatic fun hasCompletePacket(byteBuf: ByteBuf): Boolean {
            return if (byteBuf.filledSize < 37) false else byteBuf.filledSize >= byteBuf.getUInt16LE(4) + 8
        }
    }
}